src/algebra/direct_sum_graded.lean

```diff
@@ -251,105 +251,86 @@ end one
-  rw [direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply],
-  convert add_monoid_hom.congr_fun dfinsupp.sum_add_hom_single_add_hom x,
-  ext i xi : 2,
-  rw to_add_monoid_of,
+  apply add_hom_ext, intros i xi,
+  unfold has_one.one,
+  rw of_mul_of',
   exact dfinsupp.single_eq_of_sigma_eq (gmonoid.one_mul ⟨i, xi⟩),
 end
 
 private lemma mul_one (x : ⨁ i, A i) : x * 1 = x :=
+congr_at x $ show (mul_hom A).flip 1 = add_monoid_hom.id (⨁ i, A i),
 begin
-  unfold has_one.one has_mul.mul,
-  rw [to_add_monoid_of, direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply,
-    add_monoid_hom.dfinsupp_sum_add_hom_apply x _],
-  convert add_monoid_hom.congr_fun dfinsupp.sum_add_hom_single_add_hom x,
-  ext i xi : 2,
+  apply add_hom_ext, intros i xi,
+  unfold has_one.one,
+  rw [flip_apply, of_mul_of'],
   exact dfinsupp.single_eq_of_sigma_eq (gmonoid.mul_one ⟨i, xi⟩),
 end
 
--- In this proof, we use `conv` even when not necessary, because otherwise it's hard to see which
--- side the rewrite takes place on.
 private lemma mul_assoc (a b c : ⨁ i, A i) : a * b * c = a * (b * c) :=
+congr_at c $ congr_at b $ congr_at a $
+show (mul_hom A).comp_hom.comp (mul_hom A)               -- `λ a b c, a * b * c` as a bundled hom
+   = (add_monoid_hom.comp_hom flip_hom $
+       (mul_hom A).flip.comp_hom.comp (mul_hom A)).flip, -- `λ a b c, a * (b * c)` as a bundled hom
 begin
-  unfold has_one.one has_mul.mul,
-  simp only [direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply, direct_sum.of],
-  rw [←add_monoid_hom.comp_apply, dfinsupp.comp_sum_add_hom],
-
-  -- unpack `c`
-  refine add_monoid_hom.congr_fun _ c,
-  congr' 1, ext ci cx : 2,
-  rw add_monoid_hom.comp_apply,
-
-  conv_lhs {
-    rw [add_monoid_hom.dfinsupp_sum_add_hom_apply, ←add_monoid_hom.comp_apply,
-      dfinsupp.comp_sum_add_hom], },
-  conv_rhs {
-    rw [add_monoid_hom.dfinsupp_sum_add_hom_apply, ←add_monoid_hom.comp_apply,
-      dfinsupp.comp_sum_add_hom], },
-
-  -- unpack `b`
-  refine add_monoid_hom.congr_fun _ b,
-  congr' 1, ext bi bx : 2,
-
-  conv_lhs {
-    rw [add_monoid_hom.comp_apply, add_monoid_hom.dfinsupp_sum_add_hom_apply,
-      ←add_monoid_hom.comp_apply, dfinsupp.comp_sum_add_hom], },
-  conv_rhs {
-    rw [add_monoid_hom.comp_apply, add_monoid_hom.comp_apply, add_monoid_hom.comp_apply,
-      add_monoid_hom.eval_apply, add_monoid_hom.comp_hom_apply_apply, add_monoid_hom.comp_apply,
-      dfinsupp.single_add_hom_apply, dfinsupp.sum_add_hom_single,
-      add_monoid_hom.dfinsupp_sum_add_hom_apply], },
-
-  -- unpack `a`
-  refine add_monoid_hom.congr_fun _ a,
-  congr' 1, ext ai ax : 2,
-
-  conv_lhs {
-    rw [add_monoid_hom.comp_apply, add_monoid_hom.comp_apply, add_monoid_hom.comp_apply,
-      add_monoid_hom.eval_apply, add_monoid_hom.comp_hom_apply_apply, add_monoid_hom.comp_apply,
-      dfinsupp.single_add_hom_apply, dfinsupp.sum_add_hom_single], },
-
+  apply add_hom_ext, intros ai ax, apply add_hom_ext, intros bi bx, apply add_hom_ext, intros ci cx,
```

Comment:
Does `ext ai ax bi bx ci cx` do all this work for you?

Reply:
Ah, it leaves a bit of a mess. After #7011, `dsimp` ought to be able to clean it up fairly quickly though.

Reply:
I'd like that! I'm not sure how to steer `ext` in quite that direction.

If I try that (and add `: 6` to stop it from also applying `dfinsupp.ext` at the end), it produces a goal like this:
```
  ⊢ ⇑((⇑((⇑(LHS.comp (of A ai)) ax).comp (of A bi)) bx).comp (of A ci)) cx
      = ⇑((⇑((⇑(RHS.comp (of A ai)) ax).comp (of A bi)) bx).comp (of A ci)) cx
```
where for readability I've replaced the sides of the goal it started from with "LHS" and "RHS", and also eta-reduced `(λ (i : ι), A i)`. The line with `apply add_hom_ext` etc. produces this goal instead (with the same readability edits):
```
  ⊢ ⇑(⇑(⇑LHS (⇑(of A ai) ax)) (⇑(of A bi) bx)) (⇑(of A ci) cx)
      = ⇑(⇑(⇑RHS (⇑(of A ai) ax)) (⇑(of A bi) bx)) (⇑(of A ci) cx)
```

So `ext` leaves a number of extra `.comp` applications.

Using `ext?` tells me it's applying `direct_sum.add_hom_ext'` and then `add_monoid_hom.ext`. IOW it's starting with the partially-applied version of the ext lemma, and then finding the best thing to do is to fully apply the maps after all.

If we use `ext` and then unwind the `comp`s, that gets the rest of the proof to work out the same way:
```
  ext ai ax bi bx ci cx : 6,
  rw [coe_comp, function.comp_app, coe_comp, function.comp_app, coe_comp, function.comp_app,
      coe_comp, function.comp_app, coe_comp, function.comp_app, coe_comp, function.comp_app],
```
but then it's not really simpler than the `apply add_hom_ext` version.

Alternatively we could use `ext` and double down on automation using `simp only`. The whole tactic block can be shortened to this:
```
begin
  ext ai ax bi bx ci cx : 6,
  simp only [coe_comp, function.comp_app, comp_hom_apply_apply, flip_apply, flip_hom_apply,
             of_mul_of'],
  exact dfinsupp.single_eq_of_sigma_eq (gmonoid.mul_assoc ⟨ai, ax⟩ ⟨bi, bx⟩ ⟨ci, cx⟩),
end
```
Probably the main tradeoff there is that it's slower: about 630ms on my machine, vs. 180ms for the PR's current version.

Shorter still without the `only`:
```
begin
  ext ai ax bi bx ci cx : 6,
  simp [of_mul_of'],
  exact dfinsupp.single_eq_of_sigma_eq (gmonoid.mul_assoc ⟨ai, ax⟩ ⟨bi, bx⟩ ⟨ci, cx⟩),
end
```
But that's 4380ms, much too slow.

Reply:
Let's wait to see if `dsimp` provides an acceptable middle ground once the `_refl_lemma` bug is fixed.

Reply:
SGTM!

(Ah, and I just saw your second comment in this thread, after I happened to reload the page. Glad there's a diagnosed and fixable bug here.)